packages/api/tests/task.test.js

```diff
@@ -2296,6 +2296,108 @@ describe('Task tests', () => {
+
+      // Insert two products
+      const [soilAmendmentProductOne] = await mocks.productFactory(
+        { promisedFarm: [{ farm_id }] },
+        // of type 'soil_amendment_task'
+        mocks.fakeProduct({ type: 'soil_amendment_task' }),
+      );
+      const [soilAmendmentProductTwo] = await mocks.productFactory(
+        { promisedFarm: [{ farm_id }] },
+        // of type 'soil_amendment_task'
+        mocks.fakeProduct({ type: 'soil_amendment_task' }),
+      );
+
+      // Make product details available
+      await mocks.soil_amendment_productFactory({
+        promisedProduct: [soilAmendmentProductOne],
+      });
+      await mocks.soil_amendment_productFactory({
+        promisedProduct: [soilAmendmentProductTwo],
+      });
+
+      // Initial task product state
+      const soilAmendmentTaskProductData = [
+        mocks.fakeSoilAmendmentTaskProduct({
+          product_id: soilAmendmentProductOne.product_id,
+          purpose_relationships: [{ purpose_id: soilAmendmentPurpose }],
+        }),
+        mocks.fakeSoilAmendmentTaskProduct({
+          product_id: soilAmendmentProductTwo.product_id,
+          purpose_relationships: [{ purpose_id: soilAmendmentPurpose }],
+        }),
+      ];
+
+      const taskData = {
+        ...mocks.fakeTask({
+          soil_amendment_task: mocks.fakeSoilAmendmentTask({
+            method_id: soilAmendmentMethod,
+          }),
+          soil_amendment_task_products: soilAmendmentTaskProductData,
+          task_type_id,
+          owner_user_id: user_id,
+          wage_at_moment: 50,
+          assignee_user_id: user_id,
+        }),
+        locations: [{ location_id }],
+      };
+
+      // Add task
+      postTaskRequest({ user_id, farm_id }, 'soil_amendment_task', taskData, async (err, res) => {
+        expect(res.status).toBe(201);
+        const createdTask = res.body;
+        const { task_id } = createdTask;
+        // Delete abandonment reason to prevent validation error
+        delete createdTask.abandonment_reason;
+
+        // Find index of second task product
+        const indexOfFirstProduct = createdTask.soil_amendment_task_products.findIndex(
+          (tp) => tp.product_id === soilAmendmentProductOne.product_id,
+        );
+        const indexOfSecondProduct = createdTask.soil_amendment_task_products.findIndex(
+          (tp) => tp.product_id === soilAmendmentProductTwo.product_id,
+        );
+
+        // Replace second task product with new taskProduct with id of first task product
+        createdTask.soil_amendment_task_products[
+          indexOfSecondProduct
+        ] = mocks.fakeSoilAmendmentTaskProduct({
+          product_id: soilAmendmentProductOne.product_id,
+          purpose_relationships: [{ purpose_id: soilAmendmentPurpose }],
+        });
+
+        // Update first task product id to second product id
+        createdTask.soil_amendment_task_products[indexOfFirstProduct].product_id =
+          soilAmendmentProductTwo.product_id;
```

Comment:
Could you help me understand this part?
1. You created a task with two products (let's say A and B).
2. "Replace second task product (B) with new taskProduct with id of first task product (A)" - Is this different from just replacing the product_id?
3. Update first task product (A) id to second product (B) id

Now the old first task product has productB and the old second task product has product A.
I expect to have two `deleted: false` task products or two `deleted: false` task products and two `deleted: true` task products.

What am I missing?

Reply:
@SayakaOno 

2. Replacing taskProduct2 with new taskProduct essentially removes returned `id`. Simulating removing a taskProduct card in ui and adding a new one with details in one line. (There is other ways to do this). But by removing id it creates a "taskProduct 3" 

This simulates what happens on the frontend scenario where id is deleted when removed. So expectation is that taskProduct1 with id 1 updates product_id, taskProduct2 is deleted, taskProduct3 is created with original taskProduct1 product_id.

Reply:
I remembered about the `id` and came back to update my comment, but it was too late 😂 Thank you Duncan!